### ISSUE TITLE
Prompt to bring up tools inventory

### ISSUE THREAD

**Issue by tunglinwood** [snap-stanford/Biomni#7]
Hello Biomni team,

I tried a prompt to dig out all tools that Biomni have. Please check the prompt and the outcome below.

### Outcome

# Complete Biomedical Domain Tools Inventory

## Overview
This document provides a comprehensive listing of all 250+ biomedical domain tools available in the biomni agent system.

## Table of Contents
1. [Literature and Web Search Tools](#literature-and-web-search-tools) (8 tools)
2. [Database Query Tools](#database-query-tools) (31 tools)
3. [Biological Data Lake](#biological-data-lake) (100+ datasets)
4. [Software Libraries](#software-libraries) (100+ packages)
5. [Usage Examples](#usage-examples)

---

## Literature and Web Search Tools
**Module: `biomni.tool.literature`**

### 1. fetch_supplementary_info_from_doi
- **Function**: Fetches supplementary information for a paper given its DOI and returns a research log
- **Required Parameters**: 
  - `doi` (str): The paper DOI [Default: None]
- **Optional Parameters**: 
  - `output_dir` (str): Directory to save supplementary files [Default: supplementary_info]
- **Usage**: `from biomni.tool.literature import fetch_supplementary_info_from_doi`

### 2. query_arxiv
- **Function**: Query arXiv for papers based on the provided search query and return formatted results
- **Required Parameters**: 
  - `query` (str): The search query string [Default: None]
- **Optional Parameters**: 
  - `max_papers` (int): The maximum number of papers to retrieve [Default: 10]
- **Usage**: `from biomni.tool.literature import query_arxiv`

### 3. query_scholar
- **Function**: Query Google Scholar for papers based on the provided search query and return the first search result
- **Required Parameters**: 
  - `query` (str): The search query string [Default: None]
- **Usage**: `from biomni.tool.literature import query_scholar`

### 4. query_pubmed
- **Function**: Query PubMed for papers based on the provided search query
- **Required Parameters**: 
  - `query` (str): The search query string [Default: None]
- **Optional Parameters**: 
  - `max_papers` (int): Maximum number of papers [Default: 10]
  - `max_retries` (int): Maximum retry attempts [Default: 3]
- **Usage**: `from biomni.tool.literature import query_pubmed`

### 5. search_google
- **Function**: Search using Google search and return formatted results
- **Required Parameters**: 
  - `query` (str): The search query [Default: None]
- **Optional Parameters**: 
  - `num_results` (int): Number of results [Default: 3]
  - `language` (str): Language code [Default: en]
- **Usage**: `from biomni.tool.literature import search_google`

### 6. extract_url_content
- **Function**: Extract the text content of a webpage using requests and BeautifulSoup
- **Required Parameters**: 
  - `url` (str): Webpage URL to extract content from [Default: None]
- **Usage**: `from biomni.tool.literature import extract_url_content`

### 7. extract_pdf_content
- **Function**: Extract the text content of a PDF file given its URL
- **Required Parameters**: 
  - `url` (str): URL of the PDF file to extract text from [Default: None]
- **Usage**: `from biomni.tool.literature import extract_pdf_content`

### 8. advanced_web_search
- **Function**: Initiate an advanced web search by launching a specialized agent to collect relevant information and citations through multiple rounds of web searches
- **Required Parameters**: 
  - `query` (str): The search phrase you want Claude to look up [Default: None]
- **Optional Parameters**: 
  - `max_searches` (int): Upper-bound on searches [Default: 5]
  - `max_retries` (int): Maximum retry attempts [Default: 3]
- **Usage**: `from biomni.tool.literature import advanced_web_search`

---

## Database Query Tools
**Module: `biomni.tool.database`**

#### 1. query_uniprot
- **Function**: Query the UniProt REST API using either natural language or a direct endpoint
- **Required Parameters**: 
  - `prompt` (str): Natural language query about proteins [Default: None]
- **Optional Parameters**: 
  - `endpoint` (str): Direct API endpoint
  - `api_key` (str): API key for enhanced access
  - `model` (str): AI model for query processing [Default: claude-3-5-haiku-20241022]
  - `max_results` (int): Maximum results to return [Default: 5]
- **Usage**: `from biomni.tool.database import query_uniprot`

#### 2. query_alphafold
- **Function**: Query the AlphaFold Database API for protein structure predictions
- **Required Parameters**: 
  - `uniprot_id` (str): UniProt accession ID [Default: None]
- **Optional Parameters**: 
  - `endpoint` (str): Specific API endpoint
  - `residue_range` (str): Specific residue range
  - `download` (bool): Whether to download structure files
  - `output_dir` (str): Directory for downloaded files
  - `file_format` (str): Format for structure files (pdb, cif, etc.)
  - `model_version` (str): AlphaFold model version (v1-v4)
  - `model_number` (int): Specific model number
- **Usage**: `from biomni.tool.database import query_alphafold`

#### 3. query_interpro
- **Function**: Query the InterPro REST API using natural language or a direct endpoint
- **Required Parameters**: 
  - `prompt` (str): Natural language query about protein domains or families [Default: None]
- **Optional Parameters**: 
  - `endpoint` (str): Direct API endpoint
  - `api_key` (str): API key
  - `model` (str): AI model for processing
  - `max_results` (int): Maximum results [Default: 3]
- **Usage**: `from biomni.tool.database import query_interpro`

#### 4. query_pdb
- **Function**: Query the RCSB PDB database using natural language or a direct structured query
- **Required Parameters**: 
  - `prompt` (str): Natural language query about protein structures [Default: None]
- **Optional Parameters**: 
  - `query` (dict): Structured query object
  - `api_key` (str): API key
  - `model` (str): AI model for processing
  - `max_results` (int): Maximum results [Default: 3]
- **Usage**: `from biomni.tool.database import query_pdb`

#### 5. query_pdb_identifiers
- **Function**: Retrieve detailed data and/or download files for PDB identifiers
- **Required Parameters**: 
  - `identifiers` (List[str]): List of PDB identifiers [Default: None]
- **Optional Parameters**: 
  - `return_type` (str): Type of data to return [Default: entry]
  - `download` (bool): Whether to download files
  - `attributes` (List[str]): Specific attributes to retrieve
- **Usage**: `from biomni.tool.database import query_pdb_identifiers`

### Additional Database Tools (26 more)
All following similar parameter patterns with `prompt`, `endpoint`, `api_key`, `model`, and `max_results`:

- **query_kegg**: KEGG pathway and genome database
- **query_stringdb**: STRING protein-protein interaction database
- **query_iucn**: IUCN Red List species data
- **query_paleobiology**: Paleobiology database
- **query_jaspar**: JASPAR transcription factor binding profiles
- **query_worms**: World Register of Marine Species
- **query_cbioportal**: cBioPortal cancer genomics
- **query_clinvar**: ClinVar genetic variant database
- **query_geo**: Gene Expression Omnibus
- **query_dbsnp**: dbSNP genetic variation database
- **query_ucsc**: UCSC Genome Browser
- **query_ensembl**: Ensembl genome database
- **query_opentarget_genetics**: Open Targets Genetics
- **query_opentarget**: Open Targets Platform
- **query_gwas_catalog**: NHGRI-EBI GWAS Catalog
- **query_gnomad**: gnomAD population genetics
- **blast_sequence**: BLAST sequence similarity search
- **query_reactome**: Reactome pathway database
- **query_regulomedb**: RegulomeDB regulatory variants
- **query_pride**: PRIDE proteomics database
- **query_gtopdb**: Guide to Pharmacology database
- **region_to_ccre_screen**: SCREEN cis-regulatory elements
- **get_genes_near_ccre**: Genes near regulatory elements
- **query_remap**: ReMap transcription factor database
- **query_mpd**: Mouse Phenome Database
- **query_emdb**: Electron Microscopy Data Bank

---

## Biological Data Lake
**Location: `s3://biomni-datalake/`**

### Protein Interactions
- `affinity_capture-ms.parquet` - Mass spectrometry affinity capture data
- `affinity_capture-rna.parquet` - RNA affinity capture interactions
- `co-fractionation.parquet` - Co-fractionation interaction data
- `proximity_label-ms.parquet` - Proximity labeling mass spec data
- `reconstituted_complex.parquet` - Reconstituted protein complexes
- `two-hybrid.parquet` - Two-hybrid interaction data

### Drug Discovery
- `BindingDB_All_202409.tsv` - BindingDB drug binding affinities (September 2024)
- `broad_repurposing_hub_molecule_with_smiles.parquet` - Broad Institute drug repurposing molecules
- `broad_repurposing_hub_phase_moa_target_info.parquet` - Drug mechanism and target information

### Cancer Genomics (COSMIC v101, GRCh38)
- `Cosmic_Breakpoints_v101_GRCh38.csv` - Chromosomal breakpoints
- `Cosmic_CancerGeneCensus_v101_GRCh38.parquet` - Cancer Gene Census
- `Cosmic_CompleteGeneExpression_v101_GRCh38.tsv.gz` - Complete gene expression data
- `Cosmic_Fusion_v101_GRCh38.csv` - Gene fusion data
- `Cosmic_GenomeScreens_v101_GRCh38.csv` - Genome-wide screens
- `Cosmic_HallmarksOfCancer_v101_GRCh38.csv` - Cancer hallmarks
- `Cosmic_Mutant_v101_GRCh38.csv` - Mutation data
- `Cosmic_NonCodingVariants_v101_GRCh38.csv` - Non-coding variants
- `Cosmic_Resistance_v101_GRCh38.csv` - Drug resistance mutations
- `Cosmic_Sample_v101_GRCh38.csv` - Sample information
- `Cosmic_StructuralGenomeVariants_v101_GRCh38.csv` - Structural variants
- `Cosmic_Transcript_v101_GRCh38.csv` - Transcript information
- Additional COSMIC files...

### Gene Expression
- `gtex_tissue_gene_tpm.parquet` - GTEx tissue-specific gene expression (TPM)
- `GDPx2-GeneCounts.h5` - Gene count data in HDF5 format
- `GDPx2_ProcessedData_SupplementaryTable6-DifferentialExpression.csv` - Differential expression results

### Genetic Variants
- `genebass_missense_LC_filtered.pkl` - GeneBass missense variants (low confidence filtered)
- `genebass_pLoF_filtered.pkl` - GeneBass protein loss-of-function variants
- `gwas_catalog.pkl` - GWAS Catalog variants
- `variant_table.parquet` - Comprehensive variant table

### Gene Ontology and Gene Sets
- `go-plus.json` - Gene Ontology Plus annotations
- `msigdb_human_h_hallmark_geneset.parquet` - MSigDB Hallmark gene sets
- `msigdb_human_c2_curated_geneset.parquet` - MSigDB curated gene sets
- `msigdb_human_c3_motif_geneset.parquet` - MSigDB motif gene sets
- `msigdb_human_c5_ontology_geneset.parquet` - MSigDB ontology gene sets
- `msigdb_human_c6_oncogenic_geneset.parquet` - MSigDB oncogenic signatures
- `msigdb_human_c7_immunologic_geneset.parquet` - MSigDB immunologic signatures
- Additional MSigDB collections...

### MicroRNA
- `miRDB_v6.0_results.parquet` - miRDB microRNA target predictions
- `miRTarBase_microRNA_target_interaction.parquet` - miRTarBase interactions
- `miRTarBase_MicroRNA_Target_Sites.parquet` - miRNA target sites

### Disease Data
- `DisGeNET.parquet` - DisGeNET gene-disease associations
- `omim.parquet` - Online Mendelian Inheritance in Man
- `McPAS-TCR.parquet` - T-cell receptor database

### Chemical Libraries
- `enamine_cloud_library_smiles.pkl` - Enamine chemical library SMILES
- `txgnn_prediction.pkl` - TxGNN drug prediction data

---

### Bioinformatics Packages
- **biopython**: Biological computation library
- **scanpy**: Single-cell analysis in Python
- **anndata**: Annotated data structures for single-cell data
- **mudata**: Multimodal data structures
- **pysam**: Python interface for SAM/BAM files
- **pyfaidx**: FASTA file indexing and retrieval
- **pyranges**: Genomic intervals manipulation
- **pybedtools**: Python wrapper for BEDTools
- **biotite**: Computational molecular biology
- **gget**: Command line tool for genomic data retrieval
- **lifelines**: Survival analysis
- **gseapy**: Gene Set Enrichment Analysis
- **scrublet**: Doublet detection for single-cell RNA-seq
- **cellxgene-census**: Single-cell data census
- **scvelo**: RNA velocity analysis
- **harmony-pytorch**: Batch correction for single-cell data

### Cheminformatics
- **rdkit**: Chemical informatics toolkit
- **deeppurpose**: Deep learning for drug discovery
- **pyscreener**: Virtual screening toolkit
- **openbabel**: Chemical toolbox
- **descriptastorus**: Molecular descriptor calculation
- **openmm**: Molecular dynamics simulations
- **pytdc**: Therapeutics Data Commons

### Machine Learning & Statistics
- **scikit-learn**: Machine learning library
- **scipy**: Scientific computing
- **statsmodels**: Statistical modeling
- **pymc3**: Probabilistic programming
- **umap-learn**: Dimensionality reduction
- **faiss-cpu**: Similarity search and clustering
- **hyperopt**: Hyperparameter optimization

### Data Science Core
- **pandas**: Data manipulation and analysis
- **numpy**: Numerical computing
- **matplotlib**: Plotting library
- **seaborn**: Statistical data visualization
- **plotly**: Interactive plotting
- **joblib**: Parallel computing
- **tqdm**: Progress bars
- **opencv-python**: Computer vision

### Specialized Tools
- **cooler**: Hi-C data manipulation
- **trackpy**: Particle tracking
- **cellpose**: Cell segmentation
- **viennarna**: RNA secondary structure prediction
- **PyMassSpec**: Mass spectrometry analysis
- **cobra**: Constraint-based modeling
- **mageck**: CRISPR screen analysis
- **igraph**: Network analysis
- **pyscenic**: Single-cell regulatory network analysis

### R Packages (via subprocess)
- **ggplot2**: Grammar of graphics plotting
- **dplyr**: Data manipulation
- **DESeq2**: Differential gene expression analysis
- **clusterProfiler**: Functional enrichment analysis
- **Seurat**: Single-cell RNA-seq analysis
- **limma**: Linear models for microarray data
- **edgeR**: Empirical analysis of digital gene expression
- **WGCNA**: Weighted gene co-expression network analysis
- **harmony**: Batch correction
- **Matrix**: Sparse and dense matrix classes
- **readr**: Fast and friendly data import
- **stringr**: String manipulation
- **tidyr**: Data tidying

### CLI Tools (via subprocess)
- **samtools**: SAM/BAM file manipulation
- **bcftools**: VCF/BCF file manipulation
- **bedtools**: Genome arithmetic
- **tabix**: Generic indexer for TAB-delimited files
- And many more bioinformatics command-line tools

---

### Literature Search
```python
from biomni.tool.literature import query_pubmed, extract_pdf_content, advanced_web_search

# Search PubMed
papers = query_pubmed(query='CRISPR gene editing diabetes', max_papers=10)

# Extract PDF content
pdf_text = extract_pdf_content(url='https://example.com/paper.pdf')

# Advanced web search
results = advanced_web_search(query='latest COVID-19 treatments', max_searches=5)
```

### Database Queries

```python
from biomni.tool.database import query_uniprot, query_pdb, blast_sequence, query_alphafold

# Query UniProt for protein information
protein_info = query_uniprot(prompt='Find human insulin receptor')

# Search PDB for crystal structures
structures = query_pdb(prompt='Insulin receptor crystal structures with resolution better than 2A')

# BLAST sequence search
blast_result = blast_sequence(sequence='ATCGATCGATCG', database='core_nt', program='blastn')

# Get AlphaFold structure
af_structure = query_alphafold(uniprot_id='P08069', download=True, file_format='pdb')
```

### Data Lake Access

```python
import pandas as pd
import h5py

# Load GTEx expression data
gtex_data = pd.read_parquet('s3://biomni-datalake/gtex_tissue_gene_tpm.parquet')

# Load BindingDB drug data
binding_data = pd.read_csv('s3://biomni-datalake/BindingDB_All_202409.tsv', sep='       ')

# Load COSMIC cancer data
cosmic_mutations = pd.read_parquet('s3://biomni-datalake/Cosmic_CancerGeneCensus_v101_GRCh38.parquet')

# Load gene expression data (HDF5)
with h5py.File('s3://biomni-datalake/GDPx2-GeneCounts.h5', 'r') as f:
    gene_counts = f['matrix'][:]
```

### Multi-Tool Workflows

```python
# Example: Comprehensive protein analysis workflow
from biomni.tool.literature import query_pubmed
from biomni.tool.database import query_uniprot, query_alphafold, query_pdb
import pandas as pd

# Step 1: Literature search
literature = query_pubmed(query='TP53 cancer mutations', max_papers=20)

# Step 2: Get protein information
protein_info = query_uniprot(prompt='TP53 tumor suppressor protein human')

# Step 3: Get structure data
alphafold_structure = query_alphafold(uniprot_id='P04637')
pdb_structures = query_pdb(prompt='TP53 DNA binding domain crystal structures')

# Step 4: Load mutation data from data lake
cosmic_data = pd.read_parquet('s3://biomni-datalake/Cosmic_Mutant_v101_GRCh38.csv')
tp53_mutations = cosmic_data[cosmic_data['Gene'] == 'TP53']
```

---

## Summary Statistics

- **Literature Tools**: 8 functions
- **Database Tools**: 31 functions
- **Data Lake Datasets**: 100+ files
- **Python Libraries**: 100+ packages
- **R Packages**: 20+ packages
- **CLI Tools**: Multiple available
- **TOTAL**: 250+ biomedical tools and resources

**Comment by kexinhuang12345** [comment]
Hi - yes, these are all documented in biomni.stanford.edu.
